src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1419,6 +1433,150 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
+    auto shardIndex = seqToShardIndex(ledgerSeq);
+
+    if (shards_.count(shardIndex) && shards_[shardIndex].state == ShardInfo::State::finalize)
+    {
+        return shards_[shardIndex].shard->callForLedgerSQL(callback);
+    }
+
+    return false;
+}
+
+bool
+DatabaseShardImp::callForTransactionSQL(
+    LedgerIndex ledgerSeq,
+    std::function<bool(soci::session& session, LedgerIndex index)> const& callback)
+{
+    std::lock_guard lock(mutex_);
+    auto shardIndex = seqToShardIndex(ledgerSeq);
+
+    if (shards_.count(shardIndex) && shards_[shardIndex].state == ShardInfo::State::finalize)
+    {
+        return shards_[shardIndex].shard->callForTransactionSQL(callback);
+    }
+
+    return false;
+}
+
+bool
+DatabaseShardImp::iterateLedgerSQLsForward(
+    LedgerIndex minShardIndex,
+    std::function<bool(soci::session& session, LedgerIndex index)> const& callback)
+{
+    std::lock_guard lock(mutex_);
+
+    std::map<std::uint32_t, ShardInfo>::iterator it, eit;
+
+    if (minShardIndex == (LedgerIndex)-1)
+        it = shards_.begin();
+    else
+        it = shards_.lower_bound(minShardIndex);
+
+    eit = shards_.end();
+
+    for (; it != eit; it++)
+    {
+        if (it->second.state == ShardInfo::State::finalize)
+        {
+            if (!it->second.shard->callForLedgerSQL(callback))
+                return false;
+        }
+    }
+
+    return true;
+}
+
+bool
+DatabaseShardImp::iterateTransactionSQLsForward(
+    LedgerIndex minShardIndex,
+    std::function<bool(soci::session& session, LedgerIndex index)> const& callback)
+{
+    std::lock_guard lock(mutex_);
+
+    std::map<std::uint32_t, ShardInfo>::iterator it, eit;
+
+    if (minShardIndex == (LedgerIndex)-1)
+        it = shards_.begin();
+    else
+        it = shards_.lower_bound(minShardIndex);
+
+    eit = shards_.end();
+
+    for (; it != eit; it++)
+    {
+        if (it->second.state == ShardInfo::State::finalize)
+        {
+            if (!it->second.shard->callForTransactionSQL(callback))
+                return false;
+        }
+    }
+
+    return true;
+}
+
+bool
+DatabaseShardImp::iterateLedgerSQLsBack(
+    LedgerIndex maxShardIndex,
+    std::function<bool(soci::session& session, LedgerIndex index)> const& callback)
+{
+    std::lock_guard lock(mutex_);
+
+    std::map<std::uint32_t, ShardInfo>::reverse_iterator it, eit;
+
+    if (maxShardIndex == (LedgerIndex)-1)
+        it = shards_.rbegin();
+    else
+        it = std::make_reverse_iterator(shards_.upper_bound(maxShardIndex));
+
+    eit = shards_.rend();
+
+    for (; it != eit; it++)
+    {
+        if (it->second.state == ShardInfo::State::finalize)
+        {
+            if (!it->second.shard->callForLedgerSQL(callback))
+                return false;
+        }
+    }
+
+    return true;
+}
+
+bool
+DatabaseShardImp::iterateTransactionSQLsBack(
+    LedgerIndex maxShardIndex,
+    std::function<bool(soci::session& session, LedgerIndex index)> const& callback)
+{
```

Comment:
This is very similar to the function above. Maybe use a common helper function to factor out the commonality?

Reply:
Done.